geopandas/io/parquet.py

```diff
@@ -0,0 +1,223 @@
+import json
+import warnings
+
+from pandas.io.common import get_filepath_or_buffer
+from pandas.io.parquet import PyArrowImpl as PandasPyArrowImpl
+from pandas import DataFrame
+
+from shapely.wkb import loads
+
+from geopandas import GeoDataFrame, GeoSeries
+
+
+def _encode_crs(crs):
+    if isinstance(crs, str):
+        return {"proj4": crs}
+
+    return crs
+
+
+def _decode_crs(crs):
+    if crs and "proj4" in crs:
+        return crs["proj4"]
+
+    return crs
+
+
+class PyArrowImpl(PandasPyArrowImpl):
+    """Extension of Pandas PyArrowImpl to handle serializing geometries to WKB.
+    """
+
+    def write(
+        self,
+        df,
+        path,
+        compression="snappy",
+        coerce_timestamps="ms",
+        index=None,
+        **kwargs
+    ):
+        """Write data frame to a parquet file, serializing any geometry
+        columns to WKB format.
+
+        See docstring for ``to_parquet``.
+        """
+
+        self.validate_dataframe(df)
+
+        geometry_columns = df.dtypes.loc[df.dtypes == "geometry"].index.tolist()
+        geometry_column = df._geometry_column_name
+        crs = _encode_crs(df.crs)
+
+        # Convert to a DataFrame so we can convert geometries to WKB while
+        # retaining original column names
+        df = DataFrame(df.copy())
+
+        path, _, _, _ = get_filepath_or_buffer(path, mode="wb")
+
+        # Encode all geometry columns to WKB
+        for col in geometry_columns:
+            df[col] = df[col].apply(lambda g: g.wkb)
+
+        if index is None:
+            from_pandas_kwargs = {}
+        else:
+            from_pandas_kwargs = {"preserve_index": index}
+
+        table = self.api.Table.from_pandas(df, **from_pandas_kwargs)
+
+        metadata = table.schema.metadata
+
+        # Store geopandas specific metadata
+        metadata.update(
+            {
+                "geo": json.dumps(
+                    {
+                        "crs": crs,
+                        "primary": geometry_column,
+                        "columns": geometry_columns,
```

Comment:
Following pandas metadata, we could maybe add a creater library/version: 

```
    'creator': {
      'library': 'pyarrow',
      'version': '0.13.0'
    }
```

(so geopandas in this case instead of pyarrow)

Reply:
And maybe also a general "version" of the metadata (in case we change the format in the future)

Reply:
@jorisvandenbossche @darcy-r

I expanded this a little bit in the latest commits:

```
"geo": {
  "columns": ["geometry"],
  "creator": {
    "geopandas_version": "0.7.0",
    "library": "pyarrow",
    "version": "0.15.0"
  },
  "crs": {...},
  "primary_column": "geometry",
  "schema_version": "1.0.0"
}
```
(ignore `crs` for now, will be refined per discussion here)

I added `PARQUET_METADATA_VERSION` to the top of the file to allow us to set this, but it seems like the larger issue here is that we are now defining a more complete spec - and we should document that more fully - and perhaps not just in code at the top of this file.

@jorisvandenbossche @martinfleis please advise: where would it be best to document this metadata spec, in a way that allows us to introduce versioned refinements to it later?  This would also be helpful to have easy to point to for implementers that want to port this to other languages.

Reply:
I might suggest a "GeoParquet" spec definition as its own Github repo, so that ongoing format discussions could take place separate of Geopandas.